aws/build.rake

```diff
@@ -303,13 +307,17 @@ end
 
 # Update the front-end instances, in parallel, updating up to 20% of the
 # instances at any one time.
+MAX_FRONTEND_UPGRADE_FAILURES = 5
 task :deploy do
   with_hipchat_logging("deploy frontends") do
     if CDO.daemon && CDO.app_servers.any?
       Dir.chdir(deploy_dir) do
+        num_failures = 0
         thread_count = (CDO.app_servers.keys.length * 0.20).ceil
         threaded_each CDO.app_servers.keys, thread_count do |name|
-          upgrade_frontend name, CDO.app_servers[name]
+          succeeded = upgrade_frontend name, CDO.app_servers[name]
+          should_abort = !succeeded && (num_failures += 1) > MAX_FRONTEND_UPGRADE_FAILURES
```

Comment:
Nit:  Thought I wrote this way in hipchat, I would avoid the mutation in the if statement and just write this as:

if !succeeded
   num_failures += 1
   raise 'too many frontend upgrade failures, aborting deploy' if num_failures > MAX_FRONTEND_UPGRADE_FAILURES
end